mp4/lou.go

```diff
@@ -0,0 +1,279 @@
+package mp4
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/Eyevinn/mp4ff/bits"
+)
+
+// TlouBox - Track loudness info Box
+//
+// Contained in : Ludt Box (ludt)
+type TlouBox struct {
+	loudnessBaseBox
+}
+
+// TlouBox - Album loudness info Box
+//
+// Contained in : Ludt Box (ludt)
+type AlouBox struct {
+	loudnessBaseBox
+}
+
+type LoudnessBase struct {
+	EQSetID                uint8
+	DownmixID              uint8
+	DRCSetID               uint8
+	BsSamplePeakLevel      int16
+	BsTruePeakLevel        int16
+	MeasurementSystemForTP uint8
+	ReliabilityForTP       uint8
+	Measurements           []*Measurement
+}
+
+type Measurement struct {
+	MethodDefinition  uint8
+	MethodValue       uint8
+	MeasurementSystem uint8
+	Reliability       uint8
+}
+
+// loudnessBaseBox according to ISO/IEC 14496-12 Section 12.2.7.2
+type loudnessBaseBox struct {
+	Version       byte
+	Flags         uint32
+	LoudnessBases []*LoudnessBase
+}
+
+func decodeLoudnessBaseBoxSR(hdr BoxHeader, startPos uint64, sr bits.SliceReader) (*loudnessBaseBox, error) {
+	versionAndFlags := sr.ReadUint32()
+	version := byte(versionAndFlags >> 24)
+	b := &loudnessBaseBox{
+		Version: version,
+		Flags:   versionAndFlags & flagsMask,
+	}
+	var loudnessBaseCount uint8
+	if b.Version >= 1 {
+		loudnessBaseCount = 0x3f & sr.ReadUint8()
+	} else {
+		loudnessBaseCount = 1
+	}
+
+	for a := uint8(0); a < loudnessBaseCount; a++ {
+		l := &LoudnessBase{}
+		if b.Version >= 1 {
+			l.EQSetID = 0x3f & sr.ReadUint8()
+		}
+		downmixIDAndDRCSetID := sr.ReadUint16()
+		l.DownmixID = uint8(downmixIDAndDRCSetID >> 6)
+		l.DRCSetID = uint8(downmixIDAndDRCSetID & 0x3f)
+		peakLevels := sr.ReadUint24()
+		l.BsSamplePeakLevel = int16(peakLevels >> 12)
+		l.BsTruePeakLevel = int16(peakLevels & 0x0fff)
+		measurementSystemAndReliablityForTP := sr.ReadUint8()
+		l.MeasurementSystemForTP = measurementSystemAndReliablityForTP >> 4
+		l.ReliabilityForTP = measurementSystemAndReliablityForTP & 0x0f
+		measurementCount := sr.ReadUint8()
+		for i := uint8(0); i < measurementCount; i++ {
```

Comment:
Again one could preallocated the Measurements slice. In the example file there are five values, which takes a number of allocations with doubling size compared to a single one with 5.

One could also skip the pointers and have an array structs to avoid the extra layer of points. Generally arrays of small structs are more efficient than corresponding arrays of pointers to structs.